### PR TITLE
Add CombineNestedDirnameFixer

### PR DESCRIPTION
**Pull request by gharlan**
Input:

```php
dirname(dirname(dirname($path)));
```

is fixed to:

```php
dirname($path, 3);
```

(Requires PHP >= 7.0)

### CODE REVIEW

File: src/Fixer/FunctionNotation/CombineDirnameFixer.php [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+            $index = $prev;
+        }
+    }
+
+    /**
+     * @param Tokens   $tokens
+     * @param int      $index                 Index of `dirname`
+     * @param null|int $firstArgumentEndIndex Index of last token of first argument of `dirname` call
+     *
+     * @return array|bool `false` when it is not a (suppoerted) `dirname` call, an array with infos about the dirname call otherwise
```
Comment (julienfalque) [inline src/Fixer/FunctionNotation/CombineDirnameFixer.php]: Typo: *suppo**e**rted* => *supported*. Also *info* (*information*) is uncountable in english and cannot have an *s*.

File: tests/Fixer/FunctionNotation/CombineDirnameFixerTest.php [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+ * @covers \PhpCsFixer\Fixer\FunctionNotation\CombineDirnameFixer
+ */
+final class CombineDirnameFixerTest extends AbstractFixerTestCase
+{
+    /**
+     * @param string      $expected
+     * @param null|string $input
+     *
+     * @dataProvider provideFixCases
+     * @requires PHP 7.0
```
Comment (julienfalque) [inline tests/Fixer/FunctionNotation/CombineDirnameFixerTest.php]: Please add PHP <7.0 tests showing that the fixer does nothing.

Comment (gharlan) [reply tests/Fixer/FunctionNotation/CombineDirnameFixerTest.php]: done

File: src/Fixer/FunctionNotation/CombineDirnameFixer.php (hunk 1 above)
Comment (gharlan) [reply src/Fixer/FunctionNotation/CombineDirnameFixer.php]: Thanks, fixed it.

File: src/Fixer/FunctionNotation/CombineDirnameFixer.php [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        return new FixerDefinition(
+            'Replace multiple nested calls of dirname() by only one call with second level parameter. Requires PHP >= 7.0.',
+            [
+                new VersionSpecificCodeSample(
+                    "<?php\ndirname(dirname(dirname(\$path)));\n",
+                    new VersionSpecification(70000)
+                ),
+            ],
+            null,
+            'Risky when dirname() is overridden.'
```
Comment (SpacePossum) [inline src/Fixer/FunctionNotation/CombineDirnameFixer.php]: please us \`dirname\`

**Comment by gharlan** [comment]
Should this be added to the `@PHP70Migration:risky` ruleset?

**Comment by julienfalque** [comment]
Indeed, I would add it.

**Comment by SpacePossum** [comment]
@julienfalque are happy with changes?

**Comment by kubawerlos** [comment]
@gharlan please, add a test case:
```php
            [
                '<?php new dirname(dirname($path,2));',
                '<?php new dirname(dirname(dirname($path)));',
            ],
```

You may want to take a look at `FunctionsAnalyzer::isGlobalFunctionCall`.

**Comment by gharlan** [comment]
@kubawerlos good catch 👍 I will fix it.

**Comment by staabm** [comment]
Really cool. Didnt know this $level parameter of dirname.

File: README.rst [hunk 4]
```diff
@@ -390,6 +390,13 @@ Choose from the list of available rules:
 
   Calling ``unset`` on multiple items should be done in one call.
 
+* **combine_dirname** [@PHP70Migration:risky, @PHP71Migration:risky]
```
Comment (keradus) [inline README.rst]: combine dirname with... ?
`combine_multiple_dirname` or `combine_nested_dirname` ?

Comment (ntzm) [reply README.rst]: +1 for `combine_nested_dirname`

File: tests/Fixer/FunctionNotation/CombineNestedDirnameFixerTest.php [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+     *
+     * @dataProvider provideFixCases
+     * @requires PHP 7.0
+     */
+    public function testFix($expected, $input = null)
+    {
+        $this->doTest($expected, $input);
+    }
+
+    public function provideFixCases()
```
Comment (dmvdbrugge) [inline tests/Fixer/FunctionNotation/CombineNestedDirnameFixerTest.php]: What about `dirname(dirname($path, 2), 2)`? Should that result in `dirname($path, 4)`? Can that be done as well? (I'd think so...)

Comment (gharlan) [reply tests/Fixer/FunctionNotation/CombineNestedDirnameFixerTest.php]: It's already supported, I will add a test case for this.

Comment (gharlan) [reply tests/Fixer/FunctionNotation/CombineNestedDirnameFixerTest.php]: done

**Comment by keradus** [comment]
Thank you @gharlan.

**Comment by SpacePossum** [comment]
thanks @gharlan , great to see this in :)

**Pull request merged**